sdk/src/main/java/io/horizen/account/api/rpc/types/TransactionArgs.java

```diff
@@ -93,66 +93,101 @@ public TransactionArgs(
+            ));
+        }
+    }
+
     /**
      * Get sender address or use zero address if none specified.
      */
     public Address getFrom() {
         return from == null ? Address.ZERO : from;
     }
 
-    public EthereumTransaction toTransaction(NetworkParams params) throws RpcException {
+    /**
+     * Creates a new unsigned EthereumTransaction from the given arguments. If gasPrice is given and both maxFee and
+     * priorityFee are omitted a legacy transaction is created, otherwise a dynamic fee transaction (type=2) is created.
+     * Missing parameters are autofilled as follows:
+     * <ul>
+     * <li>maxPriorityFeePerGas: suggest tip cap
+     * <li>maxFeePerGas: 2 * baseFee + maxPriorityFeePerGas
+     * <li>nonce: latest nonce from mempool + 1 or current state nonce
+     * <li>gas: estimate gas
+     * </ul>
+     *
+     * @param params       to get the current chainId
+     * @param history      to calculate fee values
+     * @param pool         to get current nonce
+     * @param gasEstimator to estimate gas if required
+     * @return a new unsigned EthereumTransaction
+     * @throws RpcException If chainId is given but does not match, or if maxFeePerGas is less than
+     *                      maxPriorityFeePerGas.
+     */
+    public EthereumTransaction toTransaction(
+        NetworkParams params,
+        AccountHistory history,
+        AccountMemoryPool pool,
+        Supplier<BigInteger> gasEstimator
+    )
+        throws RpcException {
         var saneChainId = params.chainId();
         if (chainId != null && chainId.longValueExact() != saneChainId) {
             throw new RpcException(RpcError.fromCode(
                 RpcCode.InvalidParams,
                 String.format("invalid chainID: got %d, want %d", chainId, saneChainId)
             ));
         }
-        if (nonce == null) {
-            // TODO: should automatically use current nonce from "latest" state
-            throw new RpcException(RpcError.fromCode(RpcCode.InvalidParams, "missing nonce"));
-        }
-        if (gas == null) {
-            // TODO: should automatically use gas estimation
-            throw new RpcException(RpcError.fromCode(RpcCode.InvalidParams, "missing gas limit"));
+
+        var saneType = gasPrice != null
+            ? EthereumTransactionType.LegacyTxType
+            : EthereumTransactionType.DynamicFeeTxType;
+
+        if (saneType == EthereumTransactionType.DynamicFeeTxType) {
+            // if omitted use suggested tip cap
+            if (maxPriorityFeePerGas == null) maxPriorityFeePerGas = Backend.suggestTipCap(history);
+            // if omitted use 2 * baseFee + maxPriorityFeePerGas
+            if (maxFeePerGas == null) {
+                var baseFee = history.bestBlock().header().baseFee();
+                maxFeePerGas = BigInteger.TWO.multiply(baseFee).add(maxPriorityFeePerGas);
```

Comment:
we may get value from AccountState using `state.getNextBaseFee()` seems it is lighter than get the full block from History ledger.
Beside that do we need any other calculations?